lightyear/src/server/networking.rs

```diff
@@ -295,13 +320,53 @@ pub(crate) fn send(
                     metrics::counter!("transport::send::packets").increment(packets);
                     metrics::counter!("transport::send::kb").increment(bytes);
                 }
-                netserver.send(packet_byte.as_slice(), *client_id)?;
+                if let Err(e) = netserver.send(packet_byte.as_slice(), *client_id) {
+                    send_errors.push((*client_id, netserver_idx, e));
+                }
             }
             Ok(())
         })
         .unwrap_or_else(|e: ServerError| {
             error!("Error sending packets: {}", e);
         });
+
+    for (client_id, server_idx, error) in send_errors {
+        if let Some(server) = netservers.servers.get_mut(server_idx) {
+            react_to_client_error(&mut connection_manager, server, error);
+        } else {
+            error!(
+                "Could not find server for index {} when handling error",
+                server_idx
+            );
+        }
+    }
+}
+
+fn react_to_client_error(
+    connection_manager: &mut ResMut<ConnectionManager>,
+    server: &mut ServerConnection,
+    error: ConnectionError,
+) {
+    let fatal_error_client_info = match &error {
+        ConnectionError::Netcode(netcode_error) => match netcode_error {
+            NetcodeError::ClientTransport(client_id, err) => match err {
+                crate::transport::error::Error::Io(io_error) => match io_error.kind() {
+                    std::io::ErrorKind::ConnectionReset => Some((*client_id, err)),
+                    _ => None,
+                },
+                _ => None,
+            },
+            _ => None,
+        },
+        _ => None,
+    };
+    if let Some((client_id, err)) = fatal_error_client_info {
+        error!("Fatal Client Error: {:?}", error);
+        let _ = server.disconnect(client_id);
+        connection_manager.remove(client_id);
```

Comment:
Is there a better way to disconnect? `ServerCommands` is bad because then callers to `send` have to provide a `Commands`

Reply:
I'd prefer to just add `Commands` to `send` so that the disconnection logic is centralized in a single place. But yes maybe I should have `disconnect_internal` that just requires `ConnectionManager` and `ServerConnections`